ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/ParticipantPruningIT.scala

```diff
@@ -0,0 +1,601 @@
+
+        prunedEventIds = Range(itemsToPrune - transactionsPerBatch, itemsToPrune).toVector
+          .map(eventIdsLookup)
+        unprunedEventIds = Range(itemsToPrune, itemsToPrune + transactionsPerBatch).toVector
+          .map(eventIdsLookup)
+
+        prunedEventsViaFlat <- Future.sequence(
+          prunedEventIds.map(participant.flatTransactionByEventId(_, submitter).failed))
+
+        unprunedEventsViaFlat <- Future.sequence(
+          unprunedEventIds.map(participant.flatTransactionByEventId(_, submitter)))
+      } yield {
+        assert(
+          prunedEventsViaFlat.size == transactionsPerBatch,
+          "None of the pruned events via flat transactions should be visible")
+        assert(
+          unprunedEventsViaFlat.size == transactionsPerBatch,
+          "All unpruned events via flat transactions should be visible")
+        prunedEventsViaFlat.foreach(
+          assertGrpcError(_, Status.Code.NOT_FOUND, "Transaction not found, or not visible."))
+      }
+  })
+
+  test("PRPruneRepeated", "Prune succeeds when called repeatedly", allocate(SingleParty))(
+    implicit ec => {
+      case Participants(Participant(participant, submitter)) =>
+        for {
+          endOffsetAtTestStart <- participant.currentEnd()
+
+          size <- populateLedger(participant, submitter, batchesToPopulate)
+
+          txBeforePrune <- participant.transactionTrees(
+            participant
+              .getTransactionsRequest(parties = Seq(submitter), begin = endOffsetAtTestStart))
+
+          offsetToPruneUpTo = offsetAt[TransactionTree](txBeforePrune, itemsToPrune, _.offset)
+
+          offsetOfFirstSurvivingTransaction = offsetAt[TransactionTree](
+            txBeforePrune,
+            itemsToPrune + 1,
+            _.offset)
+
+          _ <- participant.prune(offsetToPruneUpTo)
+
+          txAfterPrune <- participant.transactionTrees(
+            participant
+              .getTransactionsRequest(Seq(submitter))
+              .update(_.begin := offsetToPruneUpTo))
+
+          offsetAlreadyPruned = offsetAt[TransactionTree](txBeforePrune, itemsToPrune / 2, _.offset)
+
+          _ <- participant.prune(offsetAlreadyPruned)
+
+          txAfterRedundantPrune <- participant.transactionTrees(
+            participant
+              .getTransactionsRequest(Seq(submitter))
+              .update(_.begin := offsetToPruneUpTo))
+
+          offsetToPruneUpToInSecondRealPrune = offsetAt[TransactionTree](
+            txBeforePrune,
+            itemsToPrune * 2,
+            _.offset)
+
+          offsetOfFirstSurvivingTransactionInSecondPrune = offsetAt[TransactionTree](
+            txBeforePrune,
+            itemsToPrune * 2 + 1,
+            _.offset)
+
+          _ <- participant.prune(offsetToPruneUpTo)
+
+          txAfterSecondPrune <- participant.transactionTrees(
+            participant
+              .getTransactionsRequest(Seq(submitter))
+              .update(_.begin := offsetToPruneUpToInSecondRealPrune))
```

Comment:
```suggestion
              .update(_.begin := offsetOfFirstSurvivingTransactionInSecondPrune))
```

Reply:
Good catch above - I missed that. This instance however should be `offsetToPruneUpToInSecondRealPrune` as begin is an exclusive start.